File: scala/robot-simulator/src/main/scala/Robot.scala
case class Robot(bearing: Bearing, coordinates: (Int, Int)) {
  private def x = coordinates._1
  private def y = coordinates._2

  def advance: Robot = bearing match {
    case Bearing.North => Robot(bearing, (x, y + 1))
    case Bearing.South => Robot(bearing, (x, y - 1))
    case Bearing.East => Robot(bearing, (x + 1, y))
    case Bearing.West => Robot(bearing, (x - 1, y))
  }

  def turnRight: Robot = bearing match {
    case Bearing.North => Robot(Bearing.East, coordinates)
    case Bearing.East => Robot(Bearing.South, coordinates)
    case Bearing.South => Robot(Bearing.West, coordinates)
    case Bearing.West => Robot(Bearing.North, coordinates)
  }

  def turnLeft: Robot = bearing match {
    case Bearing.North => Robot(Bearing.West, coordinates)
    case Bearing.East => Robot(Bearing.North, coordinates)
    case Bearing.South => Robot(Bearing.East, coordinates)
    case Bearing.West => Robot(Bearing.South, coordinates)
  }

  def simulate(orders: String): Robot =
    orders.foldLeft(this)((robot, op) =>
      op match {
        case 'R' => robot.turnRight
        case 'L' => robot.turnLeft
        case 'A' => robot.advance
      }
    )
}

sealed abstract class Bearing
object Bearing {
  case object North extends Bearing
  case object South extends Bearing
  case object East extends Bearing
  case object West extends Bearing
}

File: haskell/robot-simulator/src/Robot.hs
module Robot
    ( Bearing(East,North,South,West)
    , bearing
    , coordinates
    , mkRobot
    , simulate
    , turnLeft
    , turnRight
    ) where

data Bearing = North
             | East
             | South
             | West
             deriving (Eq, Show)

data Location = Location Integer Integer
  deriving (Eq, Show)

data Robot = Robot Location Bearing
  deriving (Show)

bearing :: Robot -> Bearing
bearing (Robot _ bearing) = bearing

coordinates :: Robot -> (Integer, Integer)
coordinates (Robot (Location x y) _) = (x, y)

mkRobot :: Bearing -> (Integer, Integer) -> Robot
mkRobot direction (x, y) = Robot (Location x y) direction

simulate :: Robot -> String -> Robot
simulate = foldl run

run :: Robot -> Char -> Robot
run (Robot loc heading) 'R' = Robot loc (turnRight heading)
run (Robot loc heading) 'L' = Robot loc (turnLeft heading)
run r 'A' = advance r

advance :: Robot -> Robot
advance (Robot (Location x y) North) = Robot (Location x (y+1)) North
advance (Robot (Location x y) South) = Robot (Location x (y-1)) South
advance (Robot (Location x y) East) = Robot (Location (x+1) y) East
advance (Robot (Location x y) West) = Robot (Location (x-1) y) West

turnLeft :: Bearing -> Bearing
turnLeft North = West
turnLeft East  = North
turnLeft South = East
turnLeft West  = South

turnRight :: Bearing -> Bearing
turnRight North = East
turnRight East  = South
turnRight South = West
turnRight West  = North
